parser/parser.go

```diff
@@ -61,37 +110,107 @@ var reservedIds = map[string]struct{}{
 
 // Parse converts a source input a parsed expression.
 // This function calls ParseWithMacros with AllMacros.
+//
+// Deprecated: Use NewParser().Parse() instead.
 func Parse(source common.Source) (*exprpb.ParsedExpr, *common.Errors) {
 	return ParseWithMacros(source, AllMacros)
 }
 
 // ParseWithMacros converts a source input and macros set to a parsed expression.
+//
+// Deprecated: Use NewParser().Parse() instead.
 func ParseWithMacros(source common.Source, macros []Macro) (*exprpb.ParsedExpr, *common.Errors) {
-	macroMap := make(map[string]Macro)
-	for _, m := range macros {
-		macroMap[m.MacroKey()] = m
+	return newParser(WithMacros(macros...)).Parse(source)
+}
+
+type recursionError struct {
+	message string
+}
+
+// Error implements error.
+func (re *recursionError) Error() string {
+	return re.message
+}
+
+var _ error = &recursionError{}
+
+type recursionListener struct {
+	maxDepth int
+	depth    int
+}
+
+func (rl *recursionListener) VisitTerminal(node antlr.TerminalNode) {}
+
+func (rl *recursionListener) VisitErrorNode(node antlr.ErrorNode) {}
+
+func (rl *recursionListener) EnterEveryRule(ctx antlr.ParserRuleContext) {
+	if ctx != nil && ctx.GetRuleIndex() == gen.CELParserRULE_expr {
+		if rl.depth >= rl.maxDepth {
+			rl.depth++
+			panic(&recursionError{
+				message: fmt.Sprintf("expression recursion limit exceeded: %d", rl.maxDepth),
+			})
+		}
+		rl.depth++
 	}
-	p := parser{
-		errors: &parseErrors{common.NewErrors(source)},
-		helper: newParserHelper(source),
-		macros: macroMap,
+}
+
+func (rl *recursionListener) ExitEveryRule(ctx antlr.ParserRuleContext) {
+	if ctx != nil && ctx.GetRuleIndex() == gen.CELParserRULE_expr {
+		rl.depth--
 	}
-	buf, ok := source.(runes.Buffer)
-	if !ok {
-		buf = runes.NewBuffer(source.Content())
+}
+
+var _ antlr.ParseTreeListener = &recursionListener{}
+
+type recoveryLimitError struct {
+	message string
+}
+
+// Error implements error.
+func (rl *recoveryLimitError) Error() string {
+	return rl.message
+}
+
+var _ error = &recoveryLimitError{}
+
+type recoveryLimitErrorStrategy struct {
+	*antlr.DefaultErrorStrategy
+	maxAttempts int
+	attempts    int
+}
+
+func (rl *recoveryLimitErrorStrategy) Recover(recognizer antlr.Parser, e antlr.RecognitionException) {
+	rl.checkAttempts(recognizer)
+	rl.DefaultErrorStrategy.Recover(recognizer, e)
+}
+
+func (rl *recoveryLimitErrorStrategy) RecoverInline(recognizer antlr.Parser) antlr.Token {
+	rl.checkAttempts(recognizer)
+	return rl.DefaultErrorStrategy.RecoverInline(recognizer)
+}
+
+func (rl *recoveryLimitErrorStrategy) checkAttempts(recognizer antlr.Parser) {
+	if rl.attempts >= rl.maxAttempts {
+		rl.attempts++
+		msg := fmt.Sprintf("error recovery attempt limit exceeded: %d", rl.maxAttempts)
+		recognizer.NotifyErrorListeners(msg, nil, nil)
+		panic(&recoveryLimitError{
```

Comment:
I don't think we want to panic if this happens. We just want to abort the parse step.

Reply:
I see now that I missed the custom `recover` block below. Resolving.